frontend: Create nav link for ingress

vyncent-t opened the pull request:
Related to issue #1268 

## Description
This PR addresses the enhancement requested in issue #1268 . The purpose was to transform the host in the ingress rules into a clickable link, thereby eliminating the need for users to manually copy and paste the URL to open the endpoint.

## Changes
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/c0850cd5-14fa-4dee-82c3-3117d9b3003a)

- [x]  Functional links within the field Host
- [x]  Additional links for path field for specific routing
- [x] Created logic for determining url protocols based on ingress tls, defaults to http when not using tls

## Notes
- To validate this change, I created example pod, service, and host with the host set to bing. The changes work for formatted URL stings only, but will print '*' if needed and will not print any links if no host is written.
- When using tls and host is within fields, protocol is set to https, if not using tls or if it is not found, defaults to http

Using tls
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/b9608a33-f21a-434c-b4e5-05048179a121)

Not using tls
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/3a8b2e48-c6f0-4684-81eb-993482a1f487)

joaquimrocha commented on frontend/src/components/common/LabelListItem.tsx:
```diff
@@ -6,7 +6,7 @@ export interface LabelListItemProps {
 
 export default function LabelListItem(props: LabelListItemProps) {
-  const { labels } = props;
+  const { labels = [] } = props;
```
This should be in a separate commit as it's an unrelated fix.

vyncent-t replied:
Put it in the fix branch!

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -74,7 +75,8 @@ export default function IngressDetails() {
                   label: t('Host'),
-                  getter: (data: IngressRule) => data.host || '*',
+                  getter: (data: IngressRule) =>
+                    <Link routeName={`${data.host}`}>{data.host}</Link> || '*',
```
The routeName is supposed to be a route's name from the object in router.tsx.
Also you need to validate whether the host is an actual URL before adding it as a link prop.

joaquimrocha commented on frontend/src/lib/k8s/ingress.ts:
```diff
@@ -84,7 +84,7 @@ class Ingress extends makeKubeObject<KubeIngress>('ingress') {
     const rules: IngressRule[] = [];
 
-    this.spec!.rules.forEach(({ http, host }) => {
+    this.spec!.rules?.forEach(({ http, host }) => {
```
This needs to be a separate commit.

vyncent-t replied:
Put it in the fix branch!

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
+        );
+      }
+    } catch (e) {
+      throw new Error('Invalid URL');
```
I think we should log the occurrence, though here it's just for debug purposes (console.debug("Ingress' host is not a valid URL; displaying it without a link.')) as it may fail for some reason. And it should return just the string IMO.

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -74,7 +98,7 @@ export default function IngressDetails() {
                 {
                   label: t('Host'),
-                  getter: (data: IngressRule) => data.host || '*',
+                  getter: (data: IngressRule) => isFormatURL(data.host || '*'),
```
Should we include the path too? There's something to be decided about that here.

vyncent-t replied:
Should we add the same path as the field below this?

illume replied:
I think you'd almost always want the path in the link. It doesn't seem to me very common that someone would want to visit just the host part.

vyncent-t replied:
There is a link for the host field and for the path field!

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
+      return '*';
+    }
+    try {
+      /** https://developer.mozilla.org/en-US/docs/Web/API/URL/URL
+       * new URL() constructor creates a new URL object instance only if the string passed is a valid URL
+       * if the string is not a valid URL, then an error is thrown
+       * note: adding the protocol to the url string is required for the URL constructor to work because data.host is only the host name */
```
Except for the last line (without the note prefix), I don't think the comment is needed as it's kind of the API doc.

vyncent-t replied:
Fixed it!

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
 
+  /** this function is used to format the url to be used in the Link component */
+  function isFormatURL(url: any) {
+    if (url === '*') {
+      return '*';
+    }
```
This could be removed since we'd catch it with the exception anyway, but exceptions should be avoided, so it's good! 👍

vyncent-t replied:
Fixed it!

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to format the url to be used in the Link component */
+  function isFormatURL(url: any) {
```
The name of the function seems like it returns a boolean, but it returns a component. So it should be called something like LinkOrString or something like that.

joaquimrocha reviewed (changes requested):
Needs a couple of fixes. Also, Ingresses may have an https link if they have TLS set up. So we should likely add that too. Ping me if you need more details.

illume commented:
https://kubernetes.io/docs/concepts/services-networking/ingress/

illume commented:
Missing from here is the 'prefix' information for the path. We should display this, because it's quite important... since it can change the meaning of it quite a lot.

vyncent-t commented:
> Not exactly in scope for this... but...
> 
> Missing from the ingress info is the 'prefix' information for the path. We should display this, because it's quite important... since it can change the meaning of it quite a lot.

Made a field for it!

vyncent-t commented:
> @vyncent-t as you suggested... How does wrapping work for the link? A good test for the storybook might be a very long host, and a very long path.

vyncent-t commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -36,6 +37,24 @@ export default function IngressDetails() {
+  function LinkStringFormat(url: any) {
+    try {
+      /** adding the protocol to the url string is required for the URL constructor to work because data.host is only the host name */
+      const formatURL = new URL(`http://${url}`);
```
Done!

vyncent-t commented:
May need to see if multiple paths break any links... will look into this

vyncent-t commented:
Can anyone provide context for this build error?
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/89d90aff-b547-46f2-a8d2-dcd21cc6376e)

vyncent-t commented:
Is there a way to fix this if I do not use mac?

vyncent-t commented:
> May need to see if multiple paths break any links... will look into this

Fixed this, links will now print in a list.
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/981c4745-a307-4666-9990-c8e842855c37)

illume commented:
It would be nice to have a story for the states as well.

vyncent-t commented:
WIP: this commit creates a working story using json dummydata

vyncent-t commented:
Deleted old story file that was left in

vyncent-t commented:
WIP
- Changed the backend field to list view
- Removed host from path link, links in path only render path strings but direct to full link
- Added context to path links

Need to remove comments and console logs before review

![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/e2713259-5e4f-4ceb-b379-ebfa18dc54af)

vyncent-t commented:
Commit fixes issue with rendering links that are either "*" or ""

vyncent-t commented:
Commit fixes issue with rendering links that are either "*" or ""

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+  }
+
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: any, item?: any, urlPath?: any) {
```
The item is always used, as far as I can see, so it shouldn't be optional.
And the types should be precise.

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
   }
 
+  /** this function is used to determine the protocol used in the formatURL for LinkedStringFormat */
+  function getProtocol(item: any, url: any) {
```
This function is called `getProtocol` which suggests it returns the protocol, but it returns whether https is used.

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+    if ((url === '*' || url === '' || url === undefined) && urlPath) {
+      return <Typography>{urlPath}</Typography>;
+    } else {
+      /* if the ingress does not use tls, then the url should be http */
+      if (!item.jsonData.spec?.tls) {
+        urlProtocol = 'http://';
+      } else {
+        /* if the ingress does use tls, then we compare the url to the hosts in the tls array to determine if the url should be http or https */
+        getProtocol(item, url) ? (urlProtocol = 'https://') : (urlProtocol = 'http://');
+      }
```
This logic can be re-written much more clearly if the getProtocol function returns the actual protocol.

vyncent-t replied:
I will change the name to reflect it being a Boolean return since it does not return a protocol object or string

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+  function LinkStringFormat(url: any, item?: any, urlPath?: any) {
+    let urlProtocol;
+
+    if ((url === '*' || url === '' || url === undefined) && !urlPath) {
+      return <Typography>{url}</Typography>;
+    }
+
+    if ((url === '*' || url === '' || url === undefined) && urlPath) {
+      return <Typography>{urlPath}</Typography>;
```
Can't this be rewritten as:
```typescript
if (!url || url === '*') {
  return <Typography>{urlPath || url}</Typography>;
}
```
If url is undefined, we probably want to use it as `''` instead, so:
```typescript
if (!url || url === '*') {
  return <Typography>{urlPath || url || ''}</Typography>;
}
```

vyncent-t replied:
I added notes to the relevant pieces for this to explain why one renders the host field and the other renders the path

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+        getProtocol(item, url) ? (urlProtocol = 'https://') : (urlProtocol = 'http://');
+      }
+
+      /* since we cannot access the prefix from the ingress object, we have to access it from the itemData array */
+      const itemData: any[] = item.jsonData.spec?.rules;
```
You can actually use `item.spec.rules`. Because there's a spec getter in the class.
And the itemDatas should be called rules IMO.

joaquimrocha commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+      const itemData: any[] = item.jsonData.spec?.rules;
+      let currentPathType;
+      if (itemData) {
+        itemData.forEach((rule: any) => {
+          rule.http.paths.forEach((path: any) => {
+            if (path.path === urlPath) {
+              currentPathType = path.pathType;
+            }
+          });
+        });
```
I think using `for` loops here would be more efficient in not worse case scenarios, as you can break the loop once the path is found.

joaquimrocha reviewed (dismissed):
Left a few comments in the code. Overall it's functionally fine but we can simplify the code a good deal.

vyncent-t commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+  }
+
+  function backendFormat(backend: any) {
+    const backendList: any[] = [];
```
Ah right! I had reformatted the previous backend print logic to work without the label hover

vyncent-t commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
+        });
+      }
+
+      try {
```
Changed to be based on if else on the new URL creating a valid url or undefined

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
+  }
+
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: String, item: Ingress, urlPath?: String) {
```
Components should take a `props`.

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to determine the protocol used in the formatURL for LinkedStringFormat */
```
Other comments use a different format and style.

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
+    return isHttps;
+  }
+
+  /** this function is used to format the url to be used in the Link component */
```
Other functions use a different style of commenting.

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
+    }
+  }
+
+  function backendFormat(backend: any) {
```
This one needs a comment.

vyncent-t replied:
done!

illume reviewed (changes requested):
Left a few notes.

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
+  }
+
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: String, item: Ingress, urlPath?: String) {
```
How about adding a story for LinkStringFormat? You could add the different states, which would be a nice test.

vyncent-t replied:
added!

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
+    }
+  }
+
+  function backendFormat(backend: any) {
```
This should be a React component (with a props argument).

vyncent-t commented:
added stories for refactor

vyncent-t commented:
Fixed multiple path story

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -1,12 +1,157 @@
 import SimpleTable from '../common/SimpleTable';
 
+/**
+ * This function is used to determine the protocol used in the formatURL for LinkedStringFormat
```
How about this?

```suggestion
Is https used in the ingress item?
```

"This function is used to" is a bit redundant.

illume commented on frontend/src/components/ingress/Details.tsx:
```diff
@@ -1,12 +1,157 @@
+}
+
+/**
+ * This function is used to format the url to be used in the Link component
```
Maybe get rid of this part: "This function is used to" which is a bit redundant for this component.

illume reviewed (approved):
Looks good to me. Thanks. 🎉

Nice addition of those tests.

I left a couple of minor nit picks about two comments for your consideration.

Feel free to merge it in when you're ready.
